dev/core#71 Add functionality to create PDF/Word docs from Activity searches

aydun opened the pull request:
Overview
----------------------------------------
Add a Print/Merge Document action to Activity searches

Before
----------------------------------------
Currently most other searches provide a Print/Merge action - except for Activities.

After
----------------------------------------
This adds a Print/Merge action to Activity search results

Technical Details
----------------------------------------
Activity tokens are defined but not evaluated in the legacy token system.  They are functional in the new  token system so this uses the new Civi\Token\TokenProcessor

This revises or updates classes to use the following class hierarchy:

* __base class__: `CRM_Core_Form_Task_PDFLetterCommon`
  * __defined functions__: `renderFromRows($rows, $msgPart, $formValues)`
  * __subclass__: `CRM_Contact_Form_Task_PDFLetterCommon`
    * __defined functions__: `getLoggingOptions()`, ~~`preProcess(&$form)`~~, `preProcessSingle(&$form, $cid)`, `buildQuickForm(&$form)`, `setDefaultValues()`, `formRule(...)`, `processMessageTemplate($formValues)`, `postProcess(&$form)`, `createActivities(...)`, `formatMessage(...)`, `getMimeType()`, `getTokenCategories()`
    * __consumed functions (from parent)__: *none*
    * __subclass__: `CRM_Activity_Form_Task_PDFLetterCommon`
      * __defined functions__: `postProcess(&$form)`, `postProcessActivities(&$form, $activityIds)`, `createTokenProcessor()`, `listTokens()`
      * __consumed functions (from parent)__: `processMessageTemplate()`, `renderFromRows()`
* __base class__: `HTML_Common` => `HTML_QuickForm` => `HTML_QuickForm_Page` => `CRM_Core_Form`
  * __subclass__: `CRM_Activity_Form_Task`
    * __defined functions__: `preProcess()`, `preProcessCommon(&$form)`, `setContactIDs()`, `addDefaultButtons(...)`
    * __subclass__: `CRM_Activity_Form_Task_PDF`
      * __defined functions__: ~~`preProcess()`~~, `setDefaultValues()`, `buildQuickForm()`, `postProcess()`, `listTokens()`
      * __consumed functions (from parent)__: `preProcess()`

Comments
----------------------------------------
Need to add some tests, but looking for feedback on the use of TokenProcessor first

mattwire commented:
@aydun Is there any work that could be done here to create abstract core tasks for PDF / PDFLetterCommon as I did when adding the new "Batch Update" functionality for cases in https://github.com/civicrm/civicrm-core/pull/11411 as that would help in the future to improve maintainability and possibly offer a route to migrating all PDF tasks across to the new token processor at some point in the future.

eileenmcnaughton commented on CRM/Activity/Form/Task/PDFLetterCommon.php:
```diff
@@ -0,0 +1,116 @@
+      return $html;
+    }
+
+    if (!empty($html)) {
```
I feel like this section could be shared with another class rather than duplication?  - perhaps from getRows down could be renderHtmlFromRows or similar?

aydun commented:
@eileenmcnaughton , @mattwire I've moved the rendering part to a new function in a new base class.  CRM_Contact_Form_Task_PDFLetterCommon has functioned as a base class for all the *_PDFLetterCommon so I've changed that to use the new class.  We can move truly common stuff to the new class and leave the Contact one with contact-specific code.

aydun commented:
Thanks for your comments @monishdeb and @eileen  

My thinking behind this structure is noted at https://github.com/aydun/civicrm-core/blob/51ed01419a9d4edb82d6d9ff9fdb9373bdc6376e/CRM/Core/Form/Task/PDFLetterCommon.php#L35-L40 - which is the 'alternative' at the end of @monishdeb's comments.  I would see that a later move would refactor the common parts into `CRM_Core_Form_Task_PDFLetterCommon` and then change `CRM_Activity_Form_Task_PDFLetterCommon` so that it directly extends `CRM_Core_Form_Task_PDFLetterCommon` not `CRM_Contact_Form_Task_PDFLetterCommon`

It would be good to get some thoughts from @totten about his plan for migrating to the new token processing to ensure this patch is aligned with that thinking.  In particular, https://github.com/aydun/civicrm-core/blob/51ed01419a9d4edb82d6d9ff9fdb9373bdc6376e/CRM/Activity/Tokens.php#L72-L73 seems like it could be improved.

monishdeb commented:
@eileenmcnaughton @aydun these are the specific changes I want to see in this PR and in my opinion I don't think we can treat this in a separate PR and/or issue (after we merge it) as because we are introducing some new files here and the following changes is appropriate to do it in here itself:
1. Remove ```CRM_Activity_Form_Task_PDFLetterCommon``` and use ```CRM_Core_Form_Task_PDFLetterCommon``` and move the common functions there so it inherit these in child classes.
2. Move some functions to respective classes as mentioned in the 2nd point. 
3. And re ```createTokenProcessor``` l like to get inputs from @totten 

From the feature point of view, it working perfectly 👍 So if tagging PR with ```merge ready``` implies  that it successfully implements the new task BUT need some additional changes on top of it, then should I do the same?

eileenmcnaughton commented:
@monishdeb nope - if there are required changes it's not ready for merge-ready.

Note that using traits are an option now if it fits

aydun commented:
@totten thanks for all your comments here.  As you note `processMessageTemplate()` has multiple responsibilities, not all of which are needed here, but the overhead seems fairly small and is a pre-existing problem.

One part that feels suspect is https://github.com/civicrm/civicrm-core/pull/12012/files#diff-bcf4e4fdbc2537981f5768c8b35b4164  `checkActive()` avoids unnecessary token evaluations but a list of `return` conditions doesn't seem good.

eileenmcnaughton commented:
@monishdeb can you recheck this now changes have been made to reflect input

eileenmcnaughton commented:
@monishdeb looks like this is just needing a final review from you

seamuslee001 commented:
Jenkins re test this please

aydun commented:
Lots of changes here...

@monishdeb I've changed the class hierarchy so both `CRM_Contact_Form_Task_PDFLetterCommon` and `CRM_Activity_Form_Task_PDFLetterCommon` are subclasses of `CRM_Core_Form_Task_PDFLetterCommon` with common functionality moved up to `..._Core_...`

@totten I refactored `CRM_Contact_Form_Task_PDFLetterCommon::processMessageTemplate` and pushed the template handling into the new parent class so we don't use the old token parser in `CRM_Activity_Form_Task_PDFLetterCommon` but the functionality of `CRM_Contact_Form_Task_PDFLetterCommon` should be unchanged.

This includes the `schema` change to `\Civi\Token\TokenProcessor` we discussed at the sprint.  

I've also changed `\Civi\Token\AbstractTokenSubscriber` so subclasses can override the determination of active tokens - as used by `CRM_Activity_Token` to enable variable named tokens (eg `{activity.target_N_display_name}`

totten commented on CRM/Activity/Tokens.php:
```diff
@@ -98,20 +138,122 @@ public function alterActionScheduleQuery(\Civi\ActionSchedule\Event\MailingQuery
+    }
+
+    $activities = civicrm_api3('Activity', 'get', array(
+      'id' => array('IN' => $activityIds),
```
Might fail if processing a batch of 25+ messages. Probably needs `'options' => ['limit' => 0]`?

totten replied:
Also -- the loop above (populating `$activityIds`) makes perfect sense, but let me put it under a little extra scrutiny because this patch may become an example to inform similar patches to `CRM/*/Tokens.php`. It'll probably be fairly common to extract a bunch of `activityId` values (or `contributionId` or `caseId` or whatever), and that multiplies out to several 5 SLOC bits of boilerplate. Instead, I'd vote to skip the variable `$activityIds` and have a helper function like:

```php
    $activities = civicrm_api3('Activity', 'get', array(
      'id' => array('IN' => $e->getTokenProcessor()->getContextValues('activityId')),
      'options' => ['limit' => 0]
    );
```

https://github.com/civicrm/civicrm-core/compare/master...totten:master-act-tok?expand=1#diff-a2c7ae339db570d160a69bd288516d11R171

(Aside: I've tried a few variations of `CRM_Utils_Array::collect()` and `array_map()`, but the magic bits made that more complicated. Adding `getContextValues()` seemed easier.)

aydun replied:
Thanks - I've added `getContextValues()`.  I still have `$activityIds` since it is used later in `ActivityContact.get`

totten commented on CRM/Activity/Tokens.php:
```diff
@@ -98,20 +138,122 @@ public function alterActionScheduleQuery(\Civi\ActionSchedule\Event\MailingQuery
+    }
+    else {
+      $activity = (object) $prefetch['activity'][$row->context['activityId']];
+    }
```
I'm not entirely certain where to to put this comment; maybe this works...

There's a sort of elegance in saying that `$activity` is the same thing whether it comes from the `context['actionSearchResult']` or from `$prefetch`. Intuitively, these both provide a copy of `civicrm_activity.*`, so that feels well and good.

The problem is that they're not really interchangeable. For example:

* `actionSearchResult` includes `activity_type` (the label from the OV); however, `$prefetch` comes from a default `Activity.get` API call, so it doesn't.
* `actionSearchResult` includes `case_id`; however, `$prefetch` comes from a default `Activity.get` API call, so it doesn't.
* `$prefetch` includes information about related contacts (given the `activityId`); however, `prefetch()` short-circuits in the presence of `actionSearchResult`. So the contacts are consistently prefetched.
* There's an expectation in our community that the schema for `Activity.get` API and for `civicrm_activity` are allowed to drift apart. `actionSearchResult` implicitly tracks one while `$prefetch` implicitly tracks the other.

IMHO, it would get confusing (over time) to maintain with two different functions for loading the data (`alterActionScheduleQuery()` and `prefetch()`).

Of course, `alterActionScheduleQuery()` is in there for a reason -- reducing the number of queries by tapping into the main query. We don't want to issue a new SQL query for every combination of token+recipient. Batch-loading is better.

But I've got this nagging feeling that it would be better to significantly trim down `alterActionScheduleQuery()`:
* _Change_: For `alterActionScheduleQuery()`,  we don't need a blanket `SELECT e.*` or any option-group joins.
* _Change_: For `prefetch()`, get the list of activity IDs from a mix of (a) `context['activityId']`and/or (b) `context['actionSearchResult']->entity_id` (*provided the mapping type is right*).
* _Observe_: It always loads activity data the same way.
* _Observe_: `prefetch()` only runs once for a given batch, so we're still batched-loading.
* _Observe_: When running the scheduled-reminders, the main query can be a bit thinner (easier to read/debug).
* _Observe_: If we optimize `prefetch()` a bit more (*only grab columns that are actually needed*), then that optimization applies equally well to both use-cases.

aydun replied:
Thanks for your thoughtful comments. 

I haven't touched `alterActionScheduleQuery()` or looked at that process in much detail as it was pre-existing and only used for scheduled reminders. The concept of a base SQL query that is passed around for various subscribers to modify for their own needs seems clever, and probably performant, but easy to mess up. 

As you point out the schemas for `Activity.get` API and `civicrm_activity` may diverge so tracking both could get confusing.  Which one do you prefer for these purposes?

aydun replied:
@eileenmcnaughton Thanks for the steer - I'll go that way.

aydun replied:
@eileenmcnaughton @totten I set off down the path of following `civicrm_activity` rather than `Activity.get`.  Then I got to the comments on `CRM_BAO_Activity::retrieve()` noting it should be deprecated and to use the API instead.  if we go direct to the table or via the BAO we hit the same issue of bypassing ACL's, so I've stuck with the API, at least for now.

However, by explicitly specifying the return fields, the available tokens are the ones specified in `__construct()` rather than whatever the API happens to return.

I've made the changes Tim suggested to trim down `alterActionScheduleQuery()` and move the data gathering to `prefetch()` so it is consistent between scheduled reminders and other uses.

aydun commented:
Hmm any clues what those test failures are re `Civi\API\Subscriber\WhitelistSubscriberTest::testEach` ? Doesn't seem particularly related to this PR.

eileenmcnaughton commented:
@aydun this has gotten pretty big - if you can see any smaller extractions of code moves that are non functional it might be worth breaking off a couple of smaller prepatory PRs & we can get those merged

aydun commented:
Hi @eileenmcnaughton  I've extracted 3 new PR's.  If we can get those merged I can rebase the rest.

@totten - you wanted to give `CRM_Activity_Tokens` some scrutiny as it might be used as an example for other `CRM/*/Tokens.php`  I think I have incorporated all your comments but would appreciate you looking again.

This introduces eg `{activity.target_N_display_name}`  The underscore is overloaded here and a better syntax might be `{activity.target:N:display_name}` ... but tokens historically have not allowed colons.  What are you thoughts about expanding the characters allowed in token names?

aydun commented:
FYI - this is now working on a production site.

mattwire commented:
@aydun Does this need rebasing?

mattwire commented:
@aydun I will take a look, as I require some of this functionality :-) But it may not be for a week or two!

eileenmcnaughton commented:
@aydun I'm going to close this as it's stale & not reviewable and I think that there are other related PRs that are the ones you want people to be reviewing. Good to see @mattwire is interested in helping review :-)

mattwire commented:
@aydun  All your other PRs have now been merged - do you want to rebase and re-open this one?

aydun commented:
@mattwire Yes - I've rebased locally but I don't seem to have permission to reopen this PR.  Do you have access to do that?

eileenmcnaughton commented:
@aydun doesn't look like you pushed your local rebasing to the branch...

aydun commented:
Thanks for reopening @eileenmcnaughton  There seem to be some problems when changes are force pushed to closed PR's hence wanting to get it re-opened first.  I've pushed now ... it will need squashing but I'd like to get it through the test suite first after rebasing out a bunch of other changes.

aydun commented:
Test error does not seem related.

Jenkins test this please

aydun commented:
Hmm - 2 different unrelated errors this time on the same code. Candidates for Testy Tests?

aydun commented:
This is now rebased and passing tests.  It has tests and hopefully is reviewable now - @eileenmcnaughton

aydun commented:
@eileenmcnaughton As you suggested, I have extracted the new base class into a separate PR. #13892
I'll rebase this again once that is merged.

aydun commented:
Rebased and ready for review

eileenmcnaughton commented:
@aydun I pulled out a few more changes - https://github.com/civicrm/civicrm-core/pull/13967 - we are doing some big formatting changes which would have made this bit go stale so I figured I could make the impact on this PR more positive by getting an extraction merged than by just forcing staleness

I didn't do the bit about setting properties - TBH I wasn't clear that it was saving additional queries so I left it out for now

aydun commented:
@eileenmcnaughton Many thanks for extracting out those changes.  I've rebased again to take account of that.  

I agree the properties don't save much for basic or custom tokens but it seems worthwhile for the special tokens, so for consistency I have kept that.

aydun commented:
Error looks unrelated.

Jenkins test this please.

totten commented:
(*CiviCRM Review Template WORD-1.2*)

<!-- In each category, change the word "Undecided" to "Pass" or "Issue". Add explanatory comments if prompted or desired. -->

* General standards
    * ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) __Pass__
    * ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) __Pass__
    * ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) __Undecided__: This has been a big project, so it feels like it merits some kind of documentation, although it's probably fairly clear from a user POV.
    * ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __Undecided__: Not assessed
* Developer standards
    * ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __Pass__: This is generally new code without technical implications for existing code. There are some updated event-listeners that share code-paths with other use-cases, but those have test coverage. 👍 
    * ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __Pass__: General style looks nice. 👍 I was a little concerned about the complexity of some of the new tokens, but it's pretty proportionate. The underlying functionality of `target_N_` is fairly powerful/sophisticated, so the complexity here qualifies as essential-complexity. 
    * ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __Pass__: Adds new test coverage for the new tokens.
    * ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __Undecided__: The last test run was good, but I'd suggest another test-run before merging.

eileenmcnaughton commented:
@demeritcowboy you can get out of me talking you into a unit test today by testing this if you would like :-) It has had code review so just needs r-run

demeritcowboy commented:
jenkins test this please

eileenmcnaughton commented:
@aydun looks like we moved the style goalposts - @demeritcowboy you can assume for now it's passing since it did pass before

demeritcowboy commented:
@eileenmcnaughton I really just wanted to generate a test site. I can test locally just wanted to avoid any windows weirdness.

demeritcowboy commented:
* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __Issue__
      * In the dropdown list of tokens, the word "target" is an internal word. Users would call it "With Contact". Similarly "Source" is called "Added By".      
    * [`r-doc`] __Issue__
      * The "N" part needs some documentation. But what's the use-case for example to just output "With Contact 2" _only_ from a series of selected activities? And what does 2 mean since they aren't strictly ordered in the database? The default if you leave it at "N", which the dropdown inserts verbatim, is that it picks the first one, which while arbitrary will cover the majority of activities, where there is only one. So it's still useful as-is that way.
      * Is there a way using this to just output for example "display name for all the with contacts"? (Although other than `name` wanting to concatenate a field for all with contacts might not make any sense.)
      * So it seems like the "N" part maybe isn't even necessary and maybe should just always be "first one"? Could do a fuller treatment of "N" via an extension?
    * [`r-run`] __Issue__
      * After selecting an activity and choosing the action from the actions dropdown I get `Warning: count(): Parameter must be an array or an object that implements Countable in CRM_Core_Form_Task_PDFLetterCommon::buildQuickForm() (line 150 of blah\blah\CRM\Core\Form\Task\PDFLetterCommon.php).`
      * Tokens for custom fields for the `source` don't seem to work, just displaying the token itself (e.g. `{activity.source_custom_8}`)
        * Notice: Undefined property: stdClass::$custom_8 in CRM_Activity_Tokens->evaluateToken() (line 295 of blah\blah\CRM\Activity\Tokens.php).
        * Warning: Invalid argument supplied for foreach() in Civi\Token\TokenRow->fill() (line 223 of blah\blah\Civi\Token\TokenRow.php).
     * I tested PDF. When I tried to test MS Word it came back with File Not Found. I don't think it's my windows laptop because MS Word works for the contact search print/merge function.
* Developer standards
    * [`r-tech`] __PASS__
    * [`r-code`] __PASS__
    * [`r-maint`] __PASS__ Optionally suggest adding test for source custom field as per above.
    * [`r-test`] __PASS__ Conditional pass based on previous runs.

eileenmcnaughton commented:
@aydun I'm keen to get this merged - if there are any parts that are mentioned above that can be left out & put to a follow up I'd go for that

eileenmcnaughton commented:
I have opened https://github.com/civicrm/civicrm-core/pull/14662 in attempt to get everything except the part under discussion (activity contact tokens) merged https://github.com/civicrm/civicrm-core/pull/14662

Oh I fixed

After selecting an activity and choosing the action from the actions dropdown I get Warning: count(): Parameter must be an array or an object that implements Countable in CRM_Core_Form_Task_PDFLetterCommon::buildQuickForm() (line 150 of blah\blah\CRM\Core\Form\Task\PDFLetterCommon.php).

eileenmcnaughton commented:
"Tokens for custom fields for the source don't seem to work" - they are missing from my cut down version

eileenmcnaughton commented:
On thee 3rd r-run issue - 

I tested PDF. When I tried to test MS Word it came back with File Not Found. I don't think it's my windows laptop because MS Word works for the contact search print/merge function.

- I removed the options for things other than pdf.

I think this means I have removed all the things that are not working properly from the version I've put up

mlutfy commented:
jenkins, test this please

eileenmcnaughton commented:
@aydun  looks like @mattwire has moved the rest of this too  #16200 after the current one just got merged - closing this

mattwire commented:
See https://lab.civicrm.org/mattwire/emailapi/tree/newtokenprocessor for a useful way to try it out
